freqtrade/freqai/data_kitchen.py

```diff
@@ -1027,3 +1002,78 @@ def get_unique_classes_from_labels(self, dataframe: DataFrame) -> None:
         if self.unique_classes:
             for label in self.unique_classes:
                 self.unique_class_list += list(self.unique_classes[label])
+
+# Methods called by interface.py (load_freqai_model())
+
+
+def download_all_data_for_training(timerange: TimeRange,
+                                   dp: DataProvider, config: dict) -> None:
+    """
+    Called only once upon start of bot to download the necessary data for
+    populating indicators and training the model.
+    :param timerange: TimeRange = The full data timerange for populating the indicators
+                                    and training the model.
+    :param dp: DataProvider instance attached to the strategy
+    """
+    all_pairs = copy.deepcopy(
+        config["freqai"]["feature_parameters"].get("include_corr_pairlist", [])
+    )
+    for pair in config.get("exchange", "").get("pair_whitelist"):
+        if pair not in all_pairs:
+            all_pairs.append(pair)
+
+    new_pairs_days = int((timerange.stopts - timerange.startts) / SECONDS_IN_DAY)
+    if not dp._exchange:
+        # Not realistic - this is only called in live mode.
+        raise OperationalException("Dataprovider did not have an exchange attached.")
+    refresh_backtest_ohlcv_data(
+        dp._exchange,
+        pairs=all_pairs,
+        timeframes=config["freqai"]["feature_parameters"].get("include_timeframes"),
+        datadir=config["datadir"],
+        timerange=timerange,
+        new_pairs_days=new_pairs_days,
+        erase=False,
+        data_format=config.get("dataformat_ohlcv", "json"),
+        trading_mode=config.get("trading_mode", "spot"),
+        prepend=config.get("prepend_data", False),
+    )
+
+
+def get_required_data_timerange(
+    config: dict
+) -> TimeRange:
+    """
+    Used by interface.py to pre-download necessary data for FreqAI
+    user.
+    """
+    time = datetime.datetime.now(tz=datetime.timezone.utc).timestamp()
+    trained_timerange = TimeRange()
+    data_load_timerange = TimeRange()
+
+    timeframes = config["freqai"]["feature_parameters"].get("include_timeframes")
+
+    max_tf_seconds = 0
+    for tf in timeframes:
+        secs = timeframe_to_seconds(tf)
+        if secs > max_tf_seconds:
+            max_tf_seconds = secs
+
+    max_period = config["freqai"]["feature_parameters"].get(
+        "indicator_max_period_candles", 20
+    ) * 2
+    additional_seconds = max_period * max_tf_seconds
```

Comment:
i'm ... not sure if this is correct - and i think it will result in a lot more data being downloaded than what's actually necessary.

let's assume the following include_timeframes:
`['3m',  '1d']` - as well as `indicator_max_period_candles` (how is this different to the existing `startup_candle_count`?) of 100.

The result in `additional_seconds` would be 100 * 86400 (100 days)  - which is then applied as timerange for all timeframes.
this means - we'd be downloading 100 days of 3m data - but in reality, we'd only need 100 * 60 * 3 for the 3m timeframe and 100 * 60 * 60 for the 1h timeframe - which is significantly less than the specified "additional seconds".

i therefore doubt we'll need 100 days of 3m data.

a potential solution would be to call `refresh_backtest_ohlcv_data()` in a loop of timeframes - calculating the "startup_candle_count" per timeframe instead of using the longest for everything.

Reply:
The calculation here is correct. 

`indicator_max_candles` is the max period used in the indicator calculation. Typically, this should match the maximum value of `indicator_periods_candles` list. In the case you describe, the user has set `['3m', '1d'] to `include_timeframes` and then let's say they set `[10,20]` to `indicator_periods_candles`. FreqAI loops on all their indicators, for all the `include_timeframes` for all the `indicator_periods_candles`.

So that means, the user asked for indicators to be computed for the `1d` timeframe with timewindows of 10 and 20. 

Now FreqAI needs to ensure that the first candle of training does not have a NaN, and that means ensuring that the there is a 20 day timewindow available for computing the 1d indicators with a 20 day timeperiod :-). That gets added to the user defined `train_period_days` (because we need non NaN at the beginning of the training period).  

This *needs* to be considered for all timeframes since all timeframes use all timewindows. Yes, this is where the data expansion increases quickly. I don't recommend setting 100 to your `indicator_max_period_candles` unless you actually use 100 somewhere or have it included in the `indicator_periods_candles`. 

>>indicator_max_period_candles (how is this different to the existing startup_candle_count?

They are completely different. From the doc:

```
startup_candles | Number of candles needed for backtesting only to ensure all indicators are non NaNs at the start of the first train period.  Datatype: positive integer.
-- | --
```

```
indicator_max_period_candles | The maximum period used in populate_any_indicators()  for indicator creation. FreqAI uses this information in combination  with the maximum timeframe to calculate how many data points it should  download so that the first data point does not have a NaN  Datatype: positive integer.
-- | --
```

In other words, startup_candles (used in FreqAI) is only used in backtesting, this is where the user needs to manually tell FreqAI how many candles will be needed for the full backtesting period. 

`indicator_max_period_candles` is typically the maximum of `indicator_periods_candles` but could be larger in the case where the user has a single large indicator placed outside the `indicator_periods_candles` loop.

Reply:
i honestly still don't see the difference between `startup_candles` and `indicator_max_period_candles`. 

Sure, the documentation could be changed to include freqAI for `startup_candles` - but documentation is made to cover "as is" - if usecase needs to be expanded - so be it.

Think about what `startup_candles` really does (letme rephrase it to make it's purpose clear):
"Number of candles needed to ensure that enough candles are available PRIOR to the relevant timeperiod to avoid NaN values at the beginning of the timeperiod."

for live, it's irrelevant as only the very last candle actually matters - freqai / backtesting are different - but right now, in my understanding - both serve the purpose of having no NAN candles in the dataset - hence - what would i set `indicator_max_period_candles` - vs. `startup_candles`? 

i currently can't come up with a setup where one would be 10, butt the other 100 - that'd mean that one of the 2 is misconfigured / not matching the remaining configuration.

now if both are relevant / used in freqAI is a different topic - i'm also not saying "change this now" - but i don't understand where the difference is - other than indicator_max_period_candles only applying to freqAI.

Reply:
`indicator_max_period_candles` is the highest TF candles which is used for dry/live when FreqAI automatically computes the number of "startup_candles" to ensure no NaNs. As you can see here, it is added to `train_period_days` and everything is converted to the base tf candle.

 `startup_candles` is in the base tf candle, and it includes train_period_days inside of it. The user needs to calculate this themselves. 

Is it still not clear?

Reply:
well to me it's still just another way to specify `startup_candle_count` (sorry, i meant the existing value before - as we now seem to have 3(**!**) different ways to specify the same thing) - which is also in candles (not minutes or days).

The outcome of `indicator_max_period_candles` is "i need `indicator_max_period_candles` more candles prior to the earliest date (train date in this case).
Which is exactly what `startup_candle_count` is also for.

---

Let's tackle this via an example maybe:

assuming i calculate ONLY ta.RSI(dataframe, 14)  in my populate_any_indicators, and i use 3m and 1h in my informative timeframes.

What is the correct value of `indicator_max_period_candles`, `startup_candle_count` and `startup_candles`?

Reply:
`indicator_max_period_candles` = 14. ie. the maximum period passed to talib in `populate_any_indicators()`. 

`startup_candles` can only be defined if we also know the `train_period_days`. Let's assume `train_period_days` = 10, then:

`startup_candles` = `indicator_max_period_candles` * 60 minutes/candle / 3 minutes/basecandle + `train_period_days` * 1440 minutes/day / 3 minutes/basecandle = 5080

The user doesn't need to deal with this calculation in dry/live, they only need to know "what is the maximum period I am passing to talib in `populate_any_indicators()`. It is computed for them. 

startup_candle_count is whatever the user has in the strategy.

During backtesting only, FT will set startup_candle_count += startup_candles if FreqAI is enabled.

The reason we have config['freqai']['startup_candles'] is because `startup_candle_count` is limited by the max candles allowed by exchange.  


The only change I am willing to accept here would be to add another "startup function" in the `load_freqAI_model()` function where we would compute the startup_candles automatically in the backtesting case. This was not done before because we were trying to avoid touching too much of the FT startup when we initiated development on FreqAI.

Reply:
logically (if `startup_candle_count` is assigned correctly) - it'd be 14 (since we use 14 startup candles).

> During backtesting only, FT will set startup_candle_count += startup_candles if FreqAI is enabled.

This would then get you 28 startup candles - but i don't really see why you'd need that - as your max. lookback here would be 14 - and after 14 candles, you've got all indicators filled.

---

So my initial understanding was correct - that `indicator_max_period_candles` is not needed.
`startup_candles` can (and should!) be calculated internally (it's way to complicated for the user, they'll just set it to 10.000 and hope it'll work).
`indicator_max_period_candles` is just a copy of `startup_candle_count` - which apparently i overlooked while reviewing.

It's not a problem for this PR - but something i'd like to see simplified.

Reply:
>logically (if startup_candle_count is assigned correctly) - it'd be 14 (since we use 14 startup candles)

Well no, we need 5080 startup_candles in our hypothetical scenario. What am I missing?

> it's way to complicated for the user, they'll just set it to 10.000 and hope it'll work - if not, increase some more

agreed, that was why I was writing an auto-datadownload for backtesting...but we know that also can introduce tangential problems :-). But yes, computing this automatically in load_freqai_model() is the best way to do it. 

I will add it to this PR. 

Thanks for the thorough review.

Reply:
> Well no, we need 5080 startup_candles in our hypothetical scenario. What am I missing?

This was about `indicator_max_period_candles` - not about `startup_candles`.

`startup_candles` is (to me) something with a clearly defined formula where all points are known by the bot - which means it can (and should) be calculated internally (i think we both agree on that).

---

`indicator_max_period_candles` is the one that bothers me - in the hypothetical scenario above, it'd be set to 14, too (the maximum "window" used in the indicator) - so after 14 candles, no NAN values are present anymore.
`startup_candle_count` defines the same thing - in the same way (in candles).
the only reason to have `indicator_max_period_candles` is to circumvent limitations imposed by  `startup_candle_count` - but i don't see why these can't be disabled for freqAI bots (that'd be justified - and would remove duplicate configuration of the same value).

Reply:
>the only reason to have indicator_max_period_candles is to circumvent limitations imposed by startup_candle_count 

It isn't true though, `indicator_max_period_candles` is required to know how many total startup_candles to automatically compute. The user input is the maximum period given to talib + `include_timeframes`. 

Can you answer this: how would we compute startup_candles without knowing the maximum period going into talib + the maximum timeframe in `include_timeframes` (`indicator_max_period`)?

The reason I ask you to answer this question now is because when I make the automatic startup_candles function, it will *still* require the user to input a `indicator_max_period`. Otherwise there is no way to calculate the `startup_candle`(_count).

Reply:
`startup_candle_count` needs to contain the maximum period going into talib (or any other indicator).
`indicator_max_period_candles` is the same - but for freqAI.

calculating `startup_candles` is using (based on the formula from the docs):
>  ( 4 hours * 25 max period * 60 minutes/hour + 30 day train_period_days * 1440 minutes per day ) / 5 min (base time frame) = 1488.

moving this into formula land again:
`(max_timeframes * startup_candle_count * 60 + train_period_days + 1440) / timeframe`.

if the 25 (in the sample) comes from `indicator_max_period_candles` - or from `startup_candle_count` does not matter.

Reply:
I attempted to remove `indicator_max_period_candles` and `startup_candles` in favor of keeping the standard strategy set `startup_candle_count`. You should probably review it.

Reply:
Resolving this convo to get it out of the way in the PR conversation.